java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java

```diff
@@ -0,0 +1,627 @@
+               target.validityBuffer.release();
+            }
+            target.validityBuffer = validityBuffer.slice(firstByteSource, byteSizeTarget);
+            target.validityBuffer.retain(1);
+         }
+         else {
+            /* Copy data
+             * When the first bit starts from the middle of a byte (offset != 0),
+             * copy data from src BitVector.
+             * Each byte in the target is composed by a part in i-th byte,
+             * another part in (i+1)-th byte.
+             */
+            target.allocateValidityBuffer(byteSizeTarget);
+
+            for (int i = 0; i < byteSizeTarget - 1; i++) {
+               byte b1 = getBitsFromCurrentByte(this.validityBuffer, firstByteSource + i, offset);
+               byte b2 = getBitsFromNextByte(this.validityBuffer, firstByteSource + i + 1, offset);
+
+               target.validityBuffer.setByte(i, (b1 + b2));
+            }
+
+            /* Copying the last piece is done in the following manner:
+             * if the source vector has 1 or more bytes remaining, we copy
+             * the last piece as a byte formed by shifting data
+             * from the current byte and the next byte.
+             *
+             * if the source vector has no more bytes remaining
+             * (we are at the last byte), we copy the last piece as a byte
+             * by shifting data from the current byte.
+             */
+            if((firstByteSource + byteSizeTarget - 1) < lastByteSource) {
+               byte b1 = getBitsFromCurrentByte(this.validityBuffer,
+                       firstByteSource + byteSizeTarget - 1, offset);
+               byte b2 = getBitsFromNextByte(this.validityBuffer,
+                       firstByteSource + byteSizeTarget, offset);
+
+               target.validityBuffer.setByte(byteSizeTarget - 1, b1 + b2);
+            }
+            else {
+               byte b1 = getBitsFromCurrentByte(this.validityBuffer,
+                       firstByteSource + byteSizeTarget - 1, offset);
+               target.validityBuffer.setByte(byteSizeTarget - 1, b1);
+            }
+         }
+      }
+   }
+
+   private static byte getBitsFromCurrentByte(ArrowBuf data, int index, int offset) {
+      return (byte)((data.getByte(index) & 0xFF) >>> offset);
+   }
+
+   private static byte getBitsFromNextByte(ArrowBuf data, int index, int offset) {
+      return (byte)((data.getByte(index) << (8 - offset)));
+   }
+
+
+   /******************************************************************
+    *                                                                *
+    *          common getters and setters                            *
+    *                                                                *
+    ******************************************************************/
+
+
+   /**
+    * Get the number of elements that are null in the vector
+    *
+    * @return the number of null elements.
+    */
+   public int getNullCount() {
```

Comment:
Is this function not performance critical so it's fine to recompute?

Reply:
Sorry, I don't understand. Are you saying we memoize the output of getNullCount()?

Reply:
Yeah, kind of, something like keep a `nullCount` variable in the class

Reply:
It might be surprising to user this function is not `O(1)`

Reply:
Ok, let's think about it.

Reply:
I don't think it's necessarily import for this PR to resolve this. Can you put a TODO for this?

Reply:
Sure.

Reply:
Done.